src/CarbonAware.Aggregators/src/CarbonAware/CarbonAwareAggregator.cs

```diff
@@ -48,21 +48,22 @@ public async Task<IEnumerable<EmissionsForecast>> GetCurrentForecastDataAsync(ID
     {
         using (var activity = Activity.StartActivity())
         {
-            DateTimeOffset start = GetOffsetOrDefault(props, CarbonAwareConstants.Start, DateTimeOffset.MinValue);
-            DateTimeOffset end = GetOffsetOrDefault(props, CarbonAwareConstants.End, DateTimeOffset.MaxValue);
             TimeSpan windowSize = GetDurationOrDefault(props);
             _logger.LogInformation("Aggregator getting carbon intensity forecast from data source");
 
             var forecasts = new List<EmissionsForecast>();
-            foreach(var location in GetLocationOrThrow(props))
+            foreach (var location in GetLocationOrThrow(props))
             {
                 var forecast = await this._dataSource.GetCurrentCarbonIntensityForecastAsync(location);
-                forecast.StartTime = start;
-                forecast.EndTime = end;
-                forecast.ForecastData = IntervalHelper.FilterByDuration(forecast.ForecastData, start, end);
+                var firstDataPoint = forecast.ForecastData.First();
+                var lastDataPoint = forecast.ForecastData.Last();
+                forecast.StartTime = GetOffsetOrDefault(props, CarbonAwareConstants.Start, firstDataPoint.Time);
+                forecast.EndTime = GetOffsetOrDefault(props, CarbonAwareConstants.End, lastDataPoint.Time + lastDataPoint.Duration);
+                forecast.Validate();
```

Comment:
This is "weird" imho. How come an object validates itself after being created? If I don't call validate, the object can be invalid anyways. I would expect these type of validation(s) to be external logical entities before the object is created.

Reply:
This does seem weird -- the call to the `_dataSource.GetCurrentCarbonIntensityForecastAsync` doesn't consider the start and end dates as input for the `_dataSource`, and rather only uses them to filter (and validate) the response

Reply:
well, I agree it looks to be by design (based on implementation), meaning that our `forecast/current` API is really only meant "to obtain the real-time forecast", from the (watttime) datasource

Reply:
Yes the `forecast/current` API is only meant at obtain the forecast at the current time. The historical API would get forecasts from the past.

Reply:
I believe the reason why the object validates itself is because we want the object to throw the error to the user from that call and needs to have the context of what times the forecast already has because it checks that the start and end time are valid relative to the forecast that is retrieved from WattTime.